app/src/main/java/com/kmarlow/custominstrumentation/ActivityLifecycleManager.java

```diff
@@ -102,41 +102,127 @@ public boolean createAndStartActivity(Context who, IBinder token, Intent intent)
+
+        try {
+            Class<Activity> superActivityClazz = getSuperclass(activity, ACTIVITY_PACKAGE);
+            Field finished = superActivityClazz.getDeclaredField(M_FINISHED);
+            finished.setAccessible(true);
+            boolean isFinished = finished.getBoolean(activity);
+
+            if (isFinished) {
+                return;
+            }
+
+            Field called = superActivityClazz.getDeclaredField(M_CALLED);
+            called.setAccessible(true);
+            called.setBoolean(activity, false);
+
+            instrumentation.callActivityOnPause(activity);
+
+            // FIXME: Check if super is called
+
+            Bundle outState = new Bundle();
+            Method setAllowFds = outState.getClass().getDeclaredMethod("setAllowFds", boolean.class);
+            setAllowFds.setAccessible(true);
+            setAllowFds.invoke(outState, false);
+
+            instrumentation.callActivityOnSaveInstanceState(activity, outState);
+
+            // STOP ACTIVITY
+
+            Method performStop = superActivityClazz.getDeclaredMethod("performStop");
+            performStop.setAccessible(true);
+            performStop.invoke(activity);
+
+        } catch (Exception e) {
+            Log.wtf(TAG, e);
+            Toast.makeText(activity, "Andromium is unsupported on this version", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    public void finishActivity(Activity activity) {
+        instrumentation.callActivityOnDestroy(activity);
+
+        // TODO: Determine how to GC resources
+    }
+
+    private void injectAndromiumWindow(Activity activity, Class superActivityClazz) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException, NoSuchFieldException {
         Window window = activity.getWindow();
+        AndromiumPhoneWindow21 andromiumWindow = new AndromiumPhoneWindow21(activity);
 
         if (!(window instanceof AndromiumPhoneWindow21)) {
             Log.d(TAG, "WINDOW: " + window.getClass().getName());
 
             try {
                 Field field = superActivityClazz.getDeclaredField("mWindow");
                 field.setAccessible(true);
-                field.set(activity, new AndromiumPhoneWindow21(activity));
+                field.set(activity, andromiumWindow);
             } catch (Exception error) {
                 Log.d("jesse", "GetField Activity Error: " + error);
             }
         }
+
+        // We need to re-setup the window here
+//        mWindow.setCallback(this);
+//        mWindow.setOnWindowDismissedCallback(this);
+//        mWindow.getLayoutInflater().setPrivateFactory(this);
+//        if (info.softInputMode != WindowManager.LayoutParams.SOFT_INPUT_STATE_UNSPECIFIED) {
+//            mWindow.setSoftInputMode(info.softInputMode);
+//        }
+//        if (info.uiOptions != 0) {
+//            mWindow.setUiOptions(info.uiOptions);
+//        }
+
+        andromiumWindow.setCallback(window.getCallback());
```

Comment:
This is strange, why would a compile error handle because of a missing runtime dependency? Hmm...